# Исследование надежности заемщиков


**Этот проект:** создан с целью определить какие клиенты банка более надежны как заемщики с разных перспектив. Таких как кол-во детей клиента, доход а также семейное положение

**Документация к данным:**

children — количество детей в семье

days_employed — общий трудовой стаж в днях

dob_years — возраст клиента в годах

education — уровень образования клиента

education_id — идентификатор уровня образования

family_status — семейное положение

family_status_id — идентификатор семейного положения

gender — пол клиента

income_type — тип занятости

debt — имел ли задолженность по возврату кредитов

total_income — ежемесячный доход

purpose — цель получения кредита

**План работы:**

для начала мы рассмотрим в целом данные и попробуем поверхностно понять где есть дубликаты и пропуски а также проверим более скрытые дубликаты. После чего мы разделим на категории несколько столбцов для того чтобы было легче проводить математические поерации и в целом читать данные. И последний этап, мы проведем гипотезы как регаирует кол-во детей клиента, доход а также семейное положение на возрат кредита во время

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/Users/sardortajiev/Desktop/projects/Bank_data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

In [23]:
data_pivot = data.pivot_table(index = ['children'], values = 'debt', aggfunc={'sum', 'count', 'mean'})
data_pivot

count      mean     sum
children                           
0         14091.0  0.075438  1063.0
1          4808.0  0.092346   444.0
2          2052.0  0.094542   194.0
3           330.0  0.081818    27.0
4            41.0  0.097561     4.0
5             9.0  0.000000     0.0

**Вывод:** 
Зависимость есть, кол-во детей хорошо сказывается на снижение задолжностей перед банком. Но все очень тяжело описать так как соотношение например клиентов у которых два ребенка и 5 разница в процентомном соотношении к общему кол-ву значений в столбце 0.43% на 99.57% поэтому лучше посчитать конверсию каждого случая и увидем конверсию от каждой категории.

**Вывод**

Мы можем заметить что чем больше детей тем хуже платит заемщик дажи при том что есть маленькая просадка в категории 2 и 3 ребенка особено плохо платят заемщики у которых 4 ребенка, но при этом клиенты у которых 5 детей оплачивают все во время (даже если сложно это оценить так как кол-во клиентов у которых 5 детей очень мало)

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
data.groupby('family_status')['debt'].sum()

family_status
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
гражданский брак         385
женат / замужем          927
Name: debt, dtype: int64

**Посчитаем конверсию для каждого статуса**

In [25]:
AA = (273 / len(data[data['family_status'] == 'Не женат / не замужем'])) * 100
BB = (84 / len(data[data['family_status'] == 'в разводе'])) * 100
CC = (63 / len(data[data['family_status'] == 'вдовец / вдова'])) * 100
DD = (385 / len(data[data['family_status'] == 'гражданский брак'])) * 100
EE = (927 / len(data[data['family_status'] == 'женат / замужем'])) * 100

In [26]:
print(f'Конверсия не возрата кредита по категрии семейного статуса "Не женат / не замужем" {AA}%')
print(f'Конверсия не возрата кредита по категрии семейного статуса "гражданский брак" {DD}%')
print(f'Конверсия не возрата кредита по категрии семейного статуса "женат / замужем" {EE}%')
print(f'Конверсия не возрата кредита по категрии семейного статуса "в разводе" {BB}%')
print(f'Конверсия не возрата кредита по категрии семейного статуса "вдовец / вдова" {CC}%')

Конверсия не возрата кредита по категрии семейного статуса "Не женат / не замужем" 9.763948497854077%
Конверсия не возрата кредита по категрии семейного статуса "гражданский брак" 9.313014029995163%
Конверсия не возрата кредита по категрии семейного статуса "женат / замужем" 7.560557866405676%
Конверсия не возрата кредита по категрии семейного статуса "в разводе" 7.064760302775442%
Конверсия не возрата кредита по категрии семейного статуса "вдовец / вдова" 6.624605678233439%


In [27]:
data_pivot = data.pivot_table(index = ['family_status'], values = 'debt', aggfunc={'count', 'sum', 'mean'})
data_pivot

count      mean    sum
family_status                                  
Не женат / не замужем   2796.0  0.097639  273.0
в разводе               1189.0  0.070648   84.0
вдовец / вдова           951.0  0.066246   63.0
гражданский брак        4134.0  0.093130  385.0
женат / замужем        12261.0  0.075606  927.0

**Вывод:** 
Если смотреть на конверсию и среднее значение каждой категории самый высокий риск в категории 'Не женат / не замужем' а также 'гражданский брак'. Самая не рисковая категория оказась 'вдовец / вдова'.
Чуть более рисковая категория не возрата кредита оказались категории 'в разводе' и 'женат / замужем'.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [28]:
data_pivot = data.pivot_table(index = ['total_income_category'], values = 'debt', aggfunc={'sum', 'count', 'mean'})
data_pivot

count      mean     sum
total_income_category                           
A                         25.0  0.080000     2.0
B                       5014.0  0.070602   354.0
C                      15921.0  0.084982  1353.0
D                        349.0  0.060172    21.0
E                         22.0  0.090909     2.0

Здесь будем обращать внимание только на две категории ('C' и 'B') так как у них значения значительно отличаются от остальных. 

**Вывод:** 

Касательно среднего арфиметического, здесь мы посмотрим только на два значения категрии B и C так как у них самые большие значения. Мы можем сказать что категория 'C' оплачиваю не во время и их среднее значение 0.084982 что достаточно не плохо в то время как у категории 'B' 0.070602, ниже значение.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [29]:
data.groupby('purpose_category')['debt'].sum()

purpose_category
операции с автомобилем      400
операции с недвижимостью    780
получение образования       369
проведение свадьбы          183
Name: debt, dtype: int64

In [30]:
conversion_avto = 400 / len(data[data['purpose_category'] == 'операции с автомобилем'])
conversion_avto = conversion_avto * 100
conversion_estate = 780 / len(data[data['purpose_category'] == 'операции с недвижимостью'])
conversion_estate = conversion_estate * 100
conversion_education = 369 / len(data[data['purpose_category'] == 'получение образования'])
conversion_education = conversion_education * 100
conversion_wedding = 369 / len(data[data['purpose_category'] == 'проведение свадьбы'])
conversion_wedding = conversion_wedding * 100
print(f'Конверсия риска у клиенов не возврата "проведение свадьбы" {conversion_wedding}%')
print(f'Конверсия риска у клиенов не возврата "операции с автомобилем" {conversion_avto}%')
print(f'Конверсия риска у клиенов не возврата "получение образовани" {conversion_education}%')
print(f'Конверсия риска у клиенов не возврата "операции с недвижимостью" {conversion_estate}%')

Конверсия риска у клиенов не возврата "проведение свадьбы" 15.953307392996107%
Конверсия риска у клиенов не возврата "операции с автомобилем" 9.34797849964945%
Конверсия риска у клиенов не возврата "получение образовани" 9.252758274824474%
Конверсия риска у клиенов не возврата "операции с недвижимостью" 7.255139056831923%


In [31]:
data_pivot = data.pivot_table(index = ['purpose_category'], values = 'debt', aggfunc={'count', 'mean', 'sum'})
data_pivot

count      mean    sum
purpose_category                                  
операции с автомобилем     4279.0  0.093480  400.0
операции с недвижимостью  10751.0  0.072551  780.0
получение образования      3988.0  0.092528  369.0
проведение свадьбы         2313.0  0.079118  183.0

но у меня не сходятся данные в таблице с теми что я прописывал вручную

**Вывод:**
Касательно конверсии: невооруженным глазом можно увидеть достаточно высокий риск не возрата кредита клиентов которые взяли кредит на 'проведение свадьбы' аж 16% невозрата, в то время как наименьшее кол-во невозврата у категории 'операции с недвижимостью'. 
Список конверсии людей не вернувшие кредит по категориям:

'проведение свадьбы' : 15.95%

'операции с автомобилем' : 9.35%

'получение образования' : 9.25%

'операции с недвижимостью' : 7.26%

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Мы видем пропуски в начале только в двух столбцах, это кол-во рабочих дней (стаж) и второе это заработок. Что на самом деле связанно между собой так как пропуски скорее всего из за отсутсвия работы и соответсвенно рабочего стажа. также причиной отсутствия данных может быть 
намеренное скрытие данных и технический сбой.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 
так как данные могут быть не нормального распределения и лучше всего в таких случаях брать медиану которая дает нам более приблеженное значение и более подходящее

### Шаг 4: общий вывод.

Мы провели иследование по заемщикам банка и провели три основных шага:
1. **Предобработка данных**:

Мы удалили все дубликаты и вместо путсых значение в числовых столбцах заменили их на медиану этих значеий
и также изменили тип данных. 

2. **Категоризация данных**:

Мы также раделили доходы клиентов на 5 основных категорий A, B, C, D, E

3. **Гипотезы**:

1)Есть ли зависимость между количеством детей и возвратом кредита в срок?
Клиенты у кого нет детей или кол-во детей равно трем более надежные недели чем клиенты с 1, 2, 4 детьми.

2)Есть ли зависимость между семейным положением и возвратом кредита в срок?
Менее рискованные клиенты оказались у кого есть семья но при этом кто в разводе также менее рисковые клиенты в то время как самые не надежные клиенты оказались из категории "Не женат / не замужем" и "гражданский брак"

3)Здесь мы смогли сравнить только две категории 'B' и 'C' так как у них значения гораздо больше остальных. 

4)
Достаточно высокий риск не возрата кредита клиентов которые взяли кредит на 'проведение свадьбы', а наименьшее кол-во невозврата у категории 'операции с недвижимостью' и среднее значение у категорий  'получение образования' и 'операции с автомобилем'

Спасибо большое, прочитал понял все, стало намного легче

In [32]:
data_pivot = data.pivot_table(index = ['total_income_category'], values = 'days_employed', aggfunc={'mean'}).boxplot(bins = 100
data_pivot





SyntaxError: invalid syntax (900400589.py, line 2)